src/backend/dx12/src/descriptors_cpu.rs

```diff
@@ -0,0 +1,165 @@
+pub struct HeapLinear {
+    handle_size: usize,
+    num: usize,
+    size: usize,
+    start: d3d12::D3D12_CPU_DESCRIPTOR_HANDLE,
+    raw: ComPtr<d3d12::ID3D12DescriptorHeap>,
+}
+
+impl HeapLinear {
+    pub fn new(
+        device: &ComPtr<d3d12::ID3D12Device>,
+        ty: d3d12::D3D12_DESCRIPTOR_HEAP_TYPE,
+        size: usize,
+    ) -> Self {
+        let desc = d3d12::D3D12_DESCRIPTOR_HEAP_DESC {
+            Type: ty,
+            NumDescriptors: size as u32,
+            Flags: d3d12::D3D12_DESCRIPTOR_HEAP_FLAG_NONE,
+            NodeMask: 0,
+        };
+
+        let mut heap: *mut d3d12::ID3D12DescriptorHeap = ptr::null_mut();
+        let handle_size = unsafe {
+            device.CreateDescriptorHeap(
+                &desc,
+                &d3d12::ID3D12DescriptorHeap::uuidof(),
+                &mut heap as *mut *mut _ as *mut *mut _,
+            );
+            device.GetDescriptorHandleIncrementSize(ty) as usize
+        };
+
+        let start = unsafe { (*heap).GetCPUDescriptorHandleForHeapStart() };
+
+        HeapLinear {
+            handle_size,
+            num: 0,
+            size,
+            start,
+            raw: unsafe { ComPtr::from_raw(heap) },
+        }
+    }
+
+    pub fn alloc_handle(&mut self) -> d3d12::D3D12_CPU_DESCRIPTOR_HANDLE {
+        assert!(!self.full());
+
+        let slot = self.num;
+        self.num += 1;
+
+        d3d12::D3D12_CPU_DESCRIPTOR_HANDLE {
+            ptr: self.start.ptr + self.handle_size * slot,
+        }
+    }
+
+    pub fn full(&self) -> bool {
+        self.num >= self.size
+    }
+
+    pub fn clear(&mut self) {
+        self.num = 0;
+    }
+}
+
+// Fixed-size (64) free-list allocator for CPU descriptors.
+struct Heap {
+    occupancy: u64,
+    handle_size: usize,
+    start: d3d12::D3D12_CPU_DESCRIPTOR_HANDLE,
+    raw: ComPtr<d3d12::ID3D12DescriptorHeap>,
+}
+
+impl Heap {
+    pub fn new(device: &ComPtr<d3d12::ID3D12Device>, ty: d3d12::D3D12_DESCRIPTOR_HEAP_TYPE) -> Self {
+        let desc = d3d12::D3D12_DESCRIPTOR_HEAP_DESC {
+            Type: ty,
+            NumDescriptors: 64,
+            Flags: d3d12::D3D12_DESCRIPTOR_HEAP_FLAG_NONE,
+            NodeMask: 0,
+        };
+
+        let mut heap: *mut d3d12::ID3D12DescriptorHeap = ptr::null_mut();
+        let handle_size = unsafe {
+            device.CreateDescriptorHeap(
+                &desc,
+                &d3d12::ID3D12DescriptorHeap::uuidof(),
+                &mut heap as *mut *mut _ as *mut *mut _,
+            );
+            device.GetDescriptorHandleIncrementSize(ty) as usize
+        };
+        let start = unsafe { (*heap).GetCPUDescriptorHandleForHeapStart() };
+
+        Heap {
+            handle_size,
+            occupancy: 0,
+            start,
+            raw: unsafe { ComPtr::from_raw(heap) },
+        }
+    }
+
+    pub fn alloc_handle(&mut self) -> d3d12::D3D12_CPU_DESCRIPTOR_HANDLE {
+        // Find first free slot
+        let slot = (0..64)
```

Comment:
please add TODO to use `https://doc.rust-lang.org/std/intrinsics/fn.ctlz.html (or actually do it behind the `nightly` feature ;) )